openshift-knative-operator/pkg/monitoring/monitoring.go

```diff
@@ -0,0 +1,93 @@
+package monitoring
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	mf "github.com/manifestival/manifestival"
+	"github.com/openshift-knative/serverless-operator/openshift-knative-operator/pkg/common"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/kubernetes"
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+)
+
+const (
+	DisableMonitoringEnvVar      = "DISABLE_SERVING_MONITORING_BY_DEFAULT"
+	EnableMonitoringLabel        = "openshift.io/cluster-monitoring"
+	ObservabilityCMName          = "observability"
+	ObservabilityBackendKey      = "metrics.backend-destination"
+	OpenshiftMonitoringNamespace = "openshift-monitoring"
+)
+
+func SetupServingMonitoring(api kubernetes.Interface, ks *v1alpha1.KnativeServing, log *zap.SugaredLogger) error {
+	if shouldDisableMonitoring(ks.Spec.CommonSpec.Config) {
+		log.Info("Disabling Serving monitoring")
+		if err := disableMonitoring(api, ks); err != nil {
+			return fmt.Errorf("failed to disable monitoring %w ", err)
+		}
+		return nil
+	}
+	if err := setMonitoringLabelToNamespace(ks.Namespace, api, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+func shouldDisableMonitoring(cfg v1alpha1.ConfigMapData) bool {
+	_, disableByDefault := os.LookupEnv(DisableMonitoringEnvVar)
+	backend, backendIsSet := cfg[ObservabilityCMName][ObservabilityBackendKey]
+	return (disableByDefault && !backendIsSet) || backend == "none"
+}
+
+func disableMonitoring(api kubernetes.Interface, ks *v1alpha1.KnativeServing) error {
+	// Disable metrics backend by default in case we need to eg. SRVKS-679
+	if _, ok := ks.Spec.CommonSpec.Config[ObservabilityCMName][ObservabilityBackendKey]; !ok {
+		common.Configure(&ks.Spec.CommonSpec, ObservabilityCMName, ObservabilityBackendKey, "none")
+	}
+	return setMonitoringLabelToNamespace(ks.Namespace, api, false)
+}
+
+func setMonitoringLabelToNamespace(namespace string, api kubernetes.Interface, enable bool) error {
+	ns, err := api.CoreV1().Namespaces().Get(context.Background(), namespace, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if ns.Labels == nil {
+		ns.Labels = map[string]string{}
```

Comment:
nit:

```suggestion
		ns.Labels = make(map[string]string, 1)
```

since we know we'll add one.